File: srt/srt.cpp
#include <stdio.h>
#include <assert.h>
#include <math.h>

static double get_q(double dividend, double divisor)
{
    assert(dividend/divisor < 8.0/3.0);
    assert(dividend < 4.0);

    if (dividend > 0.0)
    {
        return floor(dividend/divisor + 0.5);
    }
    else
    {
        return -floor(-dividend/divisor + 0.5);
    }
} // static double get_q(double dividend, double divisor)

static double max_dividend = 0.0;
static double max_quotient = 0.0;

static double srt(double dividend, double divisor)
{
//    printf("dividend=%f, divisor=%f\n", dividend, divisor);
    if (divisor < 0.0)
    {
        dividend = -dividend;
        divisor = -divisor;
    }

    double norm = 1.0;
    while (abs(dividend) >= 2.0)
    {
        dividend *= 0.5;
        norm *= 2.0;
    }
    while (abs(divisor) >= 2.0)
    {
        divisor *= 0.5;
        norm *= 0.5;
    }


    if (abs(dividend) > max_dividend)
        max_dividend = abs(dividend);
    if (abs(dividend/divisor) > max_quotient)
        max_quotient = abs(dividend/divisor);

    double res = 0.0;
    double factor = 1.0;
    for (int i=0; i<28; ++i)
    {
        double q = get_q(dividend, divisor);
        res += q*factor;
 //       printf(":: dividend=%f, divisor=%f, q=%f, res=%f\n", dividend, divisor, q, res);
        assert(abs(q) <= 2.0);
        factor *= 0.25;
        dividend -= q*divisor;
        dividend *= 4;
        if (abs(dividend) > max_dividend)
            max_dividend = abs(dividend);
        if (abs(dividend/divisor) > max_quotient)
            max_quotient = abs(dividend/divisor);
    }
    res *= norm;
//    printf(">> dividend=%f, divisor=%f, res=%f\n", dividend, divisor, res);
    return res;
}

static double max_diff = 0.0;

const double limit = 10000.0;

int main()
{
    for (double n=-limit; n<limit; n += 1.0)
    {
        for (double d=1.0; d<limit; d += 1.0)
        {
            double res = srt(n, d);
            double diff = abs(n/d - res);
            if (diff > max_diff)
            {
                max_diff = diff;
                printf("n={%lf}, d={%lf}, res={%lf}, diff={%lf}\n", n, d, res, diff);
            }
        }
    }

    srt(1.0, 3.0);

    printf("max_dividend={%f}\n", max_dividend);
    printf("max_quotient={%f}\n", max_quotient);
}

File: srt/srt.py
#! /usr/bin/env python3

def get_q(dividend, divisor):
    #print(f"dividend={dividend}, divisor={divisor}")
    assert abs(dividend/divisor) < 8/3
    assert abs(dividend) < 4
    if dividend > 0:
        return int(dividend/divisor + 0.5)
    else:
        return int(dividend/divisor - 0.5)

def srt(dividend, divisor):
    global max_dividend
    global max_quotient
    #print(f"dividend={dividend}, divisor={divisor}")
    if divisor < 0:
        dividend = -dividend
        divisor  = -divisor
    exp = 0
    while int(abs(dividend)) >= 2:
        dividend /= 2
        exp += 1
    while int(abs(divisor)) >= 2:
        divisor /= 2
        exp -= 1
    #print(f"=> exp={exp}")

    if abs(dividend) > max_dividend:
        max_dividend = abs(dividend)
    if abs(dividend/divisor) > max_quotient:
        max_quotient = abs(dividend/divisor)

    res = 0.0
    for i in range(28):
        q = get_q(dividend, divisor)
        #print(f"=> q={q}")
        assert abs(q) <= 2
        dividend -= q*divisor
        dividend *= 4
        res += q*(0.25**i)
        if abs(dividend) > max_dividend:
            max_dividend = abs(dividend)
        if abs(dividend/divisor) > max_quotient:
            max_quotient = abs(dividend/divisor)
    res *= 2.0**exp
    return res


def main():
    global max_dividend
    global max_quotient
    max_dividend = 0
    max_quotient = 0

    max_diff = 0
    for i in range(-1000, 1000):
        for j in range(1, 1000):
            res = srt(i, j)
            diff = abs(i/j - res)
            if diff > max_diff:
                max_diff = diff
                print(f"i={i}, j={j}, res={res}, diff={diff}")

    print(f"max_dividend={max_dividend}")
    print(f"max_quotient={max_quotient}")

main()

